fails to start in production with Duplicate method name "<init>" with signature "()V"

**Describe the bug**
I create an ApplicationScoped class twice through a separate Dependent config. It works in dev mode, but fails in production

**Expected behavior**
Either fail in dev mode or work everywhere

**Actual behavior**
fails in production with error
```
Exception in thread "main" java.lang.ExceptionInInitializerError
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstance(Unknown Source)
        at java.base/java.lang.Class.newInstance(Unknown Source)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:61)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:38)
        at io.quarkus.runtime.Quarkus.run(Quarkus.java:104)
        at io.quarkus.runner.GeneratedMain.main(GeneratedMain.zig:29)
Caused by: java.lang.RuntimeException: Failed to start quarkus
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:135)
        ... 9 more
Caused by: java.lang.RuntimeException: Failed to initialize Arc
        at io.quarkus.arc.Arc.initialize(Arc.java:26)
        at io.quarkus.arc.runtime.ArcRecorder.getContainer(ArcRecorder.java:24)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy_0(ArcProcessor$generateResources-1025303321.zig:76)
        at io.quarkus.deployment.steps.ArcProcessor$generateResources-1025303321.deploy(ArcProcessor$generateResources-1025303321.zig:40)
        at io.quarkus.runner.ApplicationImpl.<clinit>(ApplicationImpl.zig:115)
        ... 9 more
Caused by: java.lang.ClassFormatError: Duplicate method name "<init>" with signature "()V" in class file org/acme/commandmode/SomeClass
        at java.base/java.lang.ClassLoader.defineClass1(Native Method)
        at java.base/java.lang.ClassLoader.defineClass(Unknown Source)
        at java.base/java.security.SecureClassLoader.defineClass(Unknown Source)
        at java.base/jdk.internal.loader.BuiltinClassLoader.defineClass(Unknown Source)
        at java.base/jdk.internal.loader.BuiltinClassLoader.findClassOnClassPathOrNull(Unknown Source)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClassOrNull(Unknown Source)
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(Unknown Source)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
        at java.base/java.lang.Class.forName0(Native Method)
        at java.base/java.lang.Class.forName(Unknown Source)
        at org.acme.commandmode.SomeClassConfig_ProducerMethod_someClass2_3bf5c39e53a0326c125a8f6a544d05f7a852b9db_Bean.<init>(SomeClassConfig_ProducerMethod_someClass2_3bf5c39e53a0326c125a8f6a544d05f7a852b9db_Bean.zig:80)
        at io.quarkus.arc.setup.Default_ComponentsProvider.addBeans1(Default_ComponentsProvider.zig:148)
        at io.quarkus.arc.setup.Default_ComponentsProvider.getComponents(Default_ComponentsProvider.zig:38)
        at io.quarkus.arc.impl.ArcContainerImpl.<init>(ArcContainerImpl.java:113)
        at io.quarkus.arc.Arc.initialize(Arc.java:20)
        ... 13 more
```


**To Reproduce**
1. checkout https://github.com/Shohou/quarkus_bug
2. run `./gradlew clean quarkusDev` and see that `someClass2\n someClass1` message is printed
3. run `./gradlew clean build`, then `docker build --tag test -f ./docker/Dockerfile .` and `docker run -it --rm test` and see the exception

**Environment (please complete the following information):**
 - Output of `java -version`: openjdk 11.0.7 2020-04-14
 - Quarkus version or git rev: 1.9.2.Final
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): Gradle 6.5.1

**Additional context**
(Add any other context about the problem here.)

--- a/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/MultipleAddMissingNoargsConstructorTest.java
+++ b/extensions/arc/deployment/src/test/java/io/quarkus/arc/test/unproxyable/MultipleAddMissingNoargsConstructorTest.java
@@ -1,0 +1,68 @@
+package io.quarkus.arc.test.unproxyable;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.context.Dependent;
+import javax.enterprise.inject.Instance;
+import javax.inject.Inject;
+
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.test.QuarkusUnitTest;
+
+public class MultipleAddMissingNoargsConstructorTest {
+
+    @RegisterExtension
+    static final QuarkusUnitTest config = new QuarkusUnitTest()
+            .setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
+                    .addClasses(MyBean.class, MyBeanProducer.class));
+
+    @Inject
+    Instance<MyBean> myBeanInstances;
+
+    @Test
+    public void testBeansProperlyCreated() {
+        List<String> vals = new ArrayList<>(2);
+        for (MyBean myBeanInstance : myBeanInstances) {
+            vals.add(myBeanInstance.getVal());
+        }
+        Collections.sort(vals);
+        Assertions.assertEquals(Arrays.asList("val1", "val2"), vals);
+    }
+
+    static class MyBean {
+
+        private final String val;
+
+        MyBean(String val) {
+            this.val = val;
+        }
+
+        public String getVal() {
+            return val;
+        }
+    }
+
+    @Dependent
+    static class MyBeanProducer {
+
+        @ApplicationScoped
+        public MyBean myBean1() {
+            return new MyBean("val1");
+        }
+
+        @ApplicationScoped
+        public MyBean myBean2() {
+            return new MyBean("val2");
+        }
+    }
+
+}

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanDeployment.java
@@ -1066,6 +1066,7 @@ public class BeanDeployment {
             Consumer<BytecodeTransformer> bytecodeTransformerConsumer) {
 
         Map<String, List<BeanInfo>> namedBeans = new HashMap<>();
+        Set<DotName> classesReceivingNoArgsCtor = new HashSet<>();
 
         for (BeanInfo bean : beans) {
             if (bean.getName() != null) {
@@ -1076,7 +1077,7 @@ public class BeanDeployment {
                 }
                 named.add(bean);
             }
-            bean.validate(errors, validators, bytecodeTransformerConsumer);
+            bean.validate(errors, validators, bytecodeTransformerConsumer, classesReceivingNoArgsCtor);
         }
 
         if (!namedBeans.isEmpty()) {

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/BeanInfo.java
@@ -352,8 +352,8 @@ public class BeanInfo implements InjectionTargetInfo {
     }
 
     void validate(List<Throwable> errors, List<BeanDeploymentValidator> validators,
-            Consumer<BytecodeTransformer> bytecodeTransformerConsumer) {
-        Beans.validateBean(this, errors, validators, bytecodeTransformerConsumer);
+            Consumer<BytecodeTransformer> bytecodeTransformerConsumer, Set<DotName> classesReceivingNoArgsCtor) {
+        Beans.validateBean(this, errors, validators, bytecodeTransformerConsumer, classesReceivingNoArgsCtor);
     }
 
     void init(List<Throwable> errors, Consumer<BytecodeTransformer> bytecodeTransformerConsumer,

--- a/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
+++ b/independent-projects/arc/processor/src/main/java/io/quarkus/arc/processor/Beans.java
@@ -665,7 +665,7 @@ final class Beans {
     }
 
     static void validateBean(BeanInfo bean, List<Throwable> errors, List<BeanDeploymentValidator> validators,
-            Consumer<BytecodeTransformer> bytecodeTransformerConsumer) {
+            Consumer<BytecodeTransformer> bytecodeTransformerConsumer, Set<DotName> classesReceivingNoArgsCtor) {
 
         if (bean.isClassBean()) {
             ClassInfo beanClass = bean.getTarget().get().asClass();
@@ -698,9 +698,13 @@ final class Beans {
                         }
                     }
                     if (superName != null) {
-                        String superClassName = superName.toString().replace('.', '/');
-                        bytecodeTransformerConsumer.accept(new BytecodeTransformer(beanClass.name().toString(),
-                                new NoArgConstructorTransformFunction(superClassName)));
+                        if (!classesReceivingNoArgsCtor.contains(beanClass.name())) {
+                            String superClassName = superName.toString().replace('.', '/');
+                            bytecodeTransformerConsumer.accept(new BytecodeTransformer(beanClass.name().toString(),
+                                    new NoArgConstructorTransformFunction(superClassName)));
+                            classesReceivingNoArgsCtor.add(beanClass.name());
+                        }
+
                     } else {
                         errors.add(new DeploymentException(
                                 "It's not possible to add a synthetic constructor with no parameters to the unproxyable bean class: "
@@ -762,9 +766,12 @@ final class Beans {
                             }
                         }
                         if (superName != null) {
-                            String superClassName = superName.toString().replace('.', '/');
-                            bytecodeTransformerConsumer.accept(new BytecodeTransformer(returnTypeClass.name().toString(),
-                                    new NoArgConstructorTransformFunction(superClassName)));
+                            if (!classesReceivingNoArgsCtor.contains(returnTypeClass.name())) {
+                                String superClassName = superName.toString().replace('.', '/');
+                                bytecodeTransformerConsumer.accept(new BytecodeTransformer(returnTypeClass.name().toString(),
+                                        new NoArgConstructorTransformFunction(superClassName)));
+                                classesReceivingNoArgsCtor.add(returnTypeClass.name());
+                            }
                         } else {
                             errors.add(new DeploymentException(String
                                     .format("It's not possible to add a synthetic constructor with no parameters to the unproxyable return type of "
@@ -805,9 +812,12 @@ final class Beans {
                         }
                     }
                     if (superName != null) {
-                        String superClassName = superName.toString().replace('.', '/');
-                        bytecodeTransformerConsumer.accept(new BytecodeTransformer(beanClass.name().toString(),
-                                new NoArgConstructorTransformFunction(superClassName)));
+                        if (!classesReceivingNoArgsCtor.contains(beanClass.name())) {
+                            String superClassName = superName.toString().replace('.', '/');
+                            bytecodeTransformerConsumer.accept(new BytecodeTransformer(beanClass.name().toString(),
+                                    new NoArgConstructorTransformFunction(superClassName)));
+                            classesReceivingNoArgsCtor.add(beanClass.name());
+                        }
                     } else {
                         errors.add(new DeploymentException(
                                 "It's not possible to add a synthetic constructor with no parameters to the unproxyable bean class: "

